Tyler/template picker

tschie opened the pull request:
Changes template picker to use a single control/skin with model classes for groups and items.

tschie commented:
@colinanderson This is ready for review.

colinanderson commented on src/main/java/com/esri/arcgisruntime/toolkit/skins/FeatureTemplatePickerTilePaneSkin.java:
```diff
@@ -0,0 +1,329 @@
+    toggleGroup.selectedToggleProperty().addListener((observable, oldValue, newValue) ->
+      control.setSelectedFeatureTemplateItem(newValue == null ? null : (FeatureTemplateItem) newValue.getUserData())
+    );
+    control.selectedFeatureTemplateItemProperty().addListener((observable, oldValue, newValue) -> {
```
I wonder if it should be an exception to set the selected template item with an item that isn't actually a member of the toggle group i.e. an item from a different template picker.

tschie replied:
The SelectionModel javadoc seems to suggest an exception is not thrown if the item cannot be found: https://docs.oracle.com/javase/8/javafx/api/javafx/scene/control/SelectionModel.html#select-T-.

tschie replied:
Done

colinanderson commented on src/main/java/com/esri/arcgisruntime/toolkit/skins/FeatureTemplatePickerTilePaneSkin.java:
```diff
@@ -0,0 +1,329 @@
+      }
+    });
+
+    control.orientationProperty().addListener((observable, oldValue, newValue) ->
```
I think if you don't actually use `newValue` you can do these kinds of listeners as 
```control.orientationProperty().addListener(o -> updateOrientation());```

colinanderson commented on src/main/java/com/esri/arcgisruntime/toolkit/skins/FeatureTemplatePickerTilePaneSkin.java:
```diff
@@ -0,0 +1,329 @@
+
+    // update all item swatches once populated
+    swatchesNeedUpdate = true;
+    control.requestLayout();
```
Is it necessary to request layout here? I think it might work ok without that.

colinanderson commented on src/main/java/com/esri/arcgisruntime/toolkit/skins/FeatureTemplatePickerTilePaneSkin.java:
```diff
@@ -0,0 +1,329 @@
+          .toArray(CompletableFuture<?>[]::new);
+      // wait until all updates finished before updating layout normally
+      CompletableFuture.allOf(futures).thenRunAsync(() ->
+        super.layoutChildren(contentX, contentY, contentWidth, contentHeight)
```
What thread does this run on? I assume it should be running on the UI thread.

tschie replied:
Good call. I decided to call `requestLayout` here instead to ensure it's run on the JavaFX thread.

colinanderson commented on src/main/java/com/esri/arcgisruntime/toolkit/skins/FeatureTemplatePickerTilePaneSkin.java:
```diff
@@ -0,0 +1,329 @@
+  }
+
+  /**
+   * Updates the control's content pane and scroll pane based on the control's orientation.
```
Missing `@since 100.7.0`... probably missing on other methods in this class.

colinanderson commented on src/main/java/com/esri/arcgisruntime/toolkit/skins/FeatureTemplatePickerTilePaneSkin.java:
```diff
@@ -0,0 +1,329 @@
+    // add a style class so the group can be styled via CSS
+    vBox.getStyleClass().add("feature-template-group");
+    vBox.setVisible(false); // wait until the group's feature layer is loaded before displaying
+    vBox.managedProperty().bind(vBox.visibleProperty()); // hide the view when not visible
```
Is this intended - setting the managed property? I guess it means the node is not considered in the layout? My experience of unmanned layouts hasn't been great e.g. I think the callout is unmanaged.

In what circumstances do the vboxes get set to not be visible? Can the user access them to do that?

tschie replied:
Removed.

colinanderson commented on src/main/java/com/esri/arcgisruntime/toolkit/skins/FeatureTemplatePickerTilePaneSkin.java:
```diff
@@ -0,0 +1,329 @@
+    // add a tile pane to the view to display each feature template item in the group
+    TilePane tilePane = new TilePane();
+    tilePane.getStyleClass().add("tile-pane");
+    // the tile pane's orientation should be the opposite of the control's orientation
```
A "because... " would be useful here.

tschie replied:
added

colinanderson commented on src/main/java/com/esri/arcgisruntime/toolkit/skins/FeatureTemplatePickerTilePaneSkin.java:
```diff
@@ -0,0 +1,329 @@
+    });
+    vBox.getChildren().add(tilePane);
+
+    // ensure the group's feature layer is loaded to get its name for the label and its list of feature templates
```
If it isn't loaded does the template picker load it somewhere?

tschie replied:
It's loaded in FeatureTemplateGroup but it's added to the picker before load, so we need to still make sure it's loaded here. I added a load call to make it obvious it's being loaded.

colinanderson commented on src/main/java/com/esri/arcgisruntime/toolkit/skins/FeatureTemplatePickerTilePaneSkin.java:
```diff
@@ -0,0 +1,329 @@
+        imageView.setImage(swatch.get());
+        updateFuture.complete(null);
+      } catch (InterruptedException | ExecutionException | IllegalArgumentException e) {
+        updateFuture.completeExceptionally(e);
```
Does the user see this exception somewhere?

tschie replied:
I've made an update to use `CompletableFuture.allOf(futures).whenCompleteAsync` which requests a new layout whether one of the futures completes exceptionally or not. So no, the exception will not bubble up to the user.

colinanderson commented:
@tschie I don't remember seeing any property to disable templates from a feature layer that doesn't allow editing. I believe that was asked for as something that was missing from the 10.x template picker.

tschie commented on src/main/java/com/esri/arcgisruntime/toolkit/skins/feature-template-picker.css:
```diff
@@ -0,0 +1,45 @@
+}
+
+.feature-template-item:hover {
+  -fx-background-color: rgba(10, 10, 10, 0.1);
```
Ended up just removing these to use the toggle buttons default colors. The only one that might need some change is the hover state because it's pretty close to the `-fx-base` background color

tschie commented on src/integration-test/java/com/esri/arcgisruntime/toolkit/FeatureTemplatePickerTest.java:
```diff
@@ -0,0 +1,303 @@
+            });
+        });
+
+        robot.sleep(5000);
```
replaced all the sleeps with countdown latches awaiting load

tschie commented on src/integration-test/java/com/esri/arcgisruntime/toolkit/FeatureTemplatePickerTest.java:
```diff
@@ -0,0 +1,303 @@
+        FeatureTemplatePicker featureTemplatePicker = (FeatureTemplatePicker) stackPane.getChildren().get(0);
+        assertEquals(4, featureTemplatePicker.getFeatureTemplateGroups().size());
+
+        map.getOperationalLayers().forEach(layer -> robot.clickOn(layer.getName()));
```
Yes, an exception will be thrown if the name is not visible on the screen.

tschie commented on src/main/java/com/esri/arcgisruntime/toolkit/FeatureTemplateGroup.java:
```diff
@@ -0,0 +1,124 @@
+
+    featureLayer.loadAsync();
+    featureLayer.addDoneLoadingListener(() -> {
+      if (featureLayer.getLoadStatus() == LoadStatus.LOADED && featureLayer.getFeatureTable() instanceof ArcGISFeatureTable) {
```
I wasn't sure if there were feature layers which might not have feature tables before loading. Do you know?

tschie commented on src/main/java/com/esri/arcgisruntime/toolkit/FeatureTemplateItem.java:
```diff
@@ -0,0 +1,85 @@
+ */
+public final class FeatureTemplateItem {
+
+  private final ReadOnlyObjectProperty<FeatureLayer> featureLayer;
```
IntelliJ generates the getter, setter, and property methods correctly if done this way.

tschie commented on src/main/java/com/esri/arcgisruntime/toolkit/FeatureTemplatePicker.java:
```diff
@@ -16,341 +16,221 @@
- * A control which provides a view of the feature templates available in a list of feature layers and allows a template
- * to be selected. The templates can be displayed in a horizontal or vertical layout.
+ * A control which shows the feature templates available from a list of feature layers. Templates are grouped by
+ * feature layer. See skin for styling options.
```
Wasn't sure if I should link to the default skin when someone could apply their own custom skin.

tschie commented on src/main/java/com/esri/arcgisruntime/toolkit/FeatureTemplatePicker.java:
```diff
@@ -16,341 +16,221 @@
   /**
-   * Property controlling if the feature layer names are shown above that feature layer's templates. Set this to true to
-   * display the feature layer names, false otherwise.
+   * Gets the orientation of this control. See skin for effect.
```
I imagine that would depend on the skin. An accordion type skin would probably ignore orientation. This goes back to my earlier question about skin-agnostic doc: https://github.com/Esri/arcgis-runtime-toolkit-java/pull/32#discussion_r391229918.

colinanderson replied:
How about we document the current skin behaviour on the assumption that is the skin the majority of people will be using? It could say something like "with the default skin orientation does...".

tschie commented on src/main/java/com/esri/arcgisruntime/toolkit/FeatureTemplatePicker.java:
```diff
@@ -16,341 +16,221 @@
-  @Override
-  protected Skin<?> createDefaultSkin() {
-    return new FeatureTemplatePickerSkin(this);
+  public IntegerProperty symbolSizeProperty() {
```
It would be nice to have a concrete use case (example of a polyline swatch which looks better wider). That use case is also swatch-specific whereas this property applies to every swatch. I would guess that the most common use case is to have the same width & height.

colinanderson replied:
It is asked for in https://devtopia.esri.com/runtime/common-toolkit/wiki/Template-Picker
>At design time, the developer has configuration options:
>
> - to turn the names and/or descriptions on and off.
> - for the height/width/background color of the swatches.

I'm not sure what it means by design time though.

tschie replied:
But where is the use case that put it in that list?

The item and group names can be hidden via css. The background color of the toggle button could be changed via css too (on the image view).

This does not feel like min-ship to me.

tschie commented:
@colinanderson 
> I don't remember seeing any property to disable templates from a feature layer that doesn't allow editing. I believe that was asked for as something that was missing from the 10.x template picker.

Should this be a configurable property of the picker or just assumed by the skin?

I think I've addressed your comments and left some questions for you. I'll re-request a review when I've finished writing tests.

colinanderson commented:
>>I don't remember seeing any property to disable templates from a feature layer that doesn't allow editing. I believe that was asked for as something that was missing from the 10.x template picker.
>
>Should this be a configurable property of the picker or just assumed by the skin?

I think it was there as a property in the template picker before you began modifying it. If true it would disable swatches (I think actually the entire `TemplatepickerList`). It is listed in the nice to have section of the design
>to disable/grey out templates for feature layer where the create capability is not allowed.(Currently the template picker does not take account of edit capabilities. If a feature cannot be created, this is only found when you try to apply the edits).

gayl0000 commented:
@gayl0000 fyi

tschie commented:
@colinanderson 

> I think you can configure it to work from a name that ends in property.

Happy to make the change if you can point me to a project-specific way to do that.

colinanderson commented on src/main/java/com/esri/arcgisruntime/toolkit/skins/FeatureTemplatePickerTilePaneSkin.java:
```diff
@@ -289,6 +304,12 @@ private ToggleButton createToggleButtonForFeatureTemplateItem(FeatureTemplateIte
+    // disable the toggle button if its feature layer is not editable
+    toggleButton.disableProperty().bind(Bindings.createBooleanBinding(() ->
+        getSkinnable().getDisableTemplatesFromUneditableFeatureLayers() &&
+            !featureTemplateItem.getFeatureLayer().getFeatureTable().isEditable(),
```
I think maybe it should use `canAdd()` for this but you would probably need to rename the property too.

tschie replied:
@colinanderson Done. Let me know if you have a suggestion for a better property name.

colinanderson replied:
I think the name should reflect that is `canAdd` so something like `disableTemplatesWhenCanAddNotSupported` or something. I think it will be pretty difficult to come up with a short name for this.

colinanderson replied:
Maybe `disableWhenCannotAddProperty' is enough and the doc would says something along the line of "Disables templates when the feature service does not support adding new features".

tschie commented on src/test/java/com/esri/arcgisruntime/toolkit/skins/FeatureTemplatePickerTilePaneSkinUnitTest.java:
```diff
@@ -0,0 +1,73 @@
+    try {
+      Platform.startup(() -> {});
+    } catch (Exception ex) {
+      // toolkit already initialized
```
@colinanderson it shouldn't fail though. I do this because when multiple tests classes are run, the first one will need to run Platform.startup but the others don't. I can't think of a better way to do that.

tschie commented on src/main/java/com/esri/arcgisruntime/toolkit/skins/FeatureTemplatePickerTilePaneSkin.java:
```diff
@@ -0,0 +1,338 @@
+      // wait until all updates finished before updating layout normally
+      CompletableFuture.allOf(futures).whenCompleteAsync((future, exception) -> getSkinnable().requestLayout());
+    } else {
+      super.layoutChildren(contentX, contentY, contentWidth, contentHeight);
```
Doesn't seem like it was needed

tschie commented:
@colinanderson Thanks. I think I've addressed your comments.

colinanderson commented on src/test/java/com/esri/arcgisruntime/toolkit/util/PlatformUtils.java:
```diff
@@ -0,0 +1,23 @@
+package com.esri.arcgisruntime.toolkit.util;
```
Missing copyright.
